File: src/frontend/src/types/config/memoryBackend.ts
/**
 * Memory backend configuration types.
 *
 * CrewAI's unified Memory class replaces the legacy short/long/entity split
 * with a single scoped memory store, so these types no longer carry per-tier
 * enable flags or per-tier index/table names.
 */

export enum MemoryBackendType {
  DEFAULT = 'default', // Kasal unified Memory (local SQLite)
  // Legacy value kept so saved configurations still parse. Memory runs on
  // Lakebase or the local store; a databricks config falls back to local.
  DATABRICKS = 'databricks',
  LAKEBASE = 'lakebase', // Lakebase pgvector
}

/**
 * Memory Tuning knobs for a teamspace's memory. Every field here reaches the
 * layer that uses it — the scoring ones the storage backend, the rest the
 * engine's `Memory` — so a value set in the panel is never silently dropped.
 */
export interface MemoryTuningConfig {
  // Composite score weights (semantic + keyword + recency + importance —
  // should roughly sum to 1.0).
  semantic_weight?: number;
  keyword_weight?: number;
  recency_weight?: number;
  importance_weight?: number;
  recency_half_life_days?: number;
  /** Minimum semantic similarity for a memory to be recalled at all (default 0.35). */
  relevance_threshold?: number;
  /**
   * Blended-score floor below which a recall returns nothing. Unset = the
   * embedder's own default (0.75 Databricks, 0.62 local Ollama fallback).
   */
  recall_min_score?: number;

  // Consolidation.
  consolidation_threshold?: number;
  consolidation_limit?: number;
  default_importance?: number;

  // Recall depth control.
  confidence_threshold_high?: number;
  confidence_threshold_low?: number;
  complex_query_threshold?: number;
  exploration_budget?: number;
  query_analysis_threshold?: number;

  // LLM override for memory analysis.
  memory_llm_model?: string;
}

export interface DatabricksMemoryConfig {
  // Memory endpoint (Direct Access for dynamic record-level writes).
  endpoint_name: string;

  // Memory index — one index for every MemoryRecord.
  memory_index: string;

  // Document search endpoint + index (unrelated to memory).
  document_endpoint_name?: string;
  document_index?: string;

  // Database configuration
  catalog?: string;
  schema?: string;

  // Authentication (optional - can use environment variables)
  workspace_url?: string;
  auth_type?: 'default' | 'pat' | 'service_principal';

  // For PAT authentication
  personal_access_token?: string;

  // For Service Principal authentication
  service_principal_client_id?: string;
  service_principal_client_secret?: string;

  // Vector configuration
  embedding_dimension?: number;
}

export interface LakebaseMemoryConfig {
  instance_name?: string;
  embedding_dimension?: number;
  // Memory table — one table for every MemoryRecord.
  memory_table?: string;
  tables_initialized?: boolean;
}

export interface MemoryBackendConfig {
  backend_type: MemoryBackendType;

  // Backend-specific configuration
  databricks_config?: DatabricksMemoryConfig;
  lakebase_config?: LakebaseMemoryConfig;

  // Recall-scoring parameters. The wire field name is fixed by the API.
  cognitive_config?: MemoryTuningConfig;

  // Database persistence fields
  is_default?: boolean;
  is_active?: boolean;

  // Escape hatch for experimental backend-specific options
  custom_config?: Record<string, unknown>;
}

// Default configurations for easy setup
export const DEFAULT_MEMORY_BACKEND_CONFIG: MemoryBackendConfig = {
  backend_type: MemoryBackendType.DEFAULT,
};

export const DEFAULT_DATABRICKS_CONFIG: DatabricksMemoryConfig = {
  endpoint_name: '',
  memory_index: '',
  embedding_dimension: 1024,
  auth_type: 'default',
};

export const DEFAULT_LAKEBASE_CONFIG: LakebaseMemoryConfig = {
  embedding_dimension: 1024,
  memory_table: 'crew_memory',
  tables_initialized: false,
};

/**
 * The EFFECTIVE defaults — what the backend applies when a knob is unset —
 * so an untouched slider shows the value that is actually in force
 * (local_storage_backend.LocalMemoryStorage and engine/memory.Memory).
 */
export const MEMORY_TUNING_DEFAULTS: Required<
  Pick<
    MemoryTuningConfig,
    | 'semantic_weight'
    | 'keyword_weight'
    | 'recency_weight'
    | 'importance_weight'
    | 'recency_half_life_days'
    | 'relevance_threshold'
    | 'recall_min_score'
    | 'consolidation_threshold'
    | 'consolidation_limit'
    | 'default_importance'
    | 'confidence_threshold_high'
    | 'confidence_threshold_low'
    | 'complex_query_threshold'
    | 'exploration_budget'
    | 'query_analysis_threshold'
  >
> = {
  semantic_weight: 0.6,
  keyword_weight: 0.15,
  recency_weight: 0.15,
  importance_weight: 0.1,
  recency_half_life_days: 30,
  relevance_threshold: 0.35,
  recall_min_score: 0.75,
  consolidation_threshold: 0.85,
  consolidation_limit: 5,
  default_importance: 0.5,
  confidence_threshold_high: 0.8,
  confidence_threshold_low: 0.5,
  complex_query_threshold: 0.7,
  exploration_budget: 1,
  query_analysis_threshold: 200,
};

// Validation helpers
export const isValidMemoryBackendConfig = (config: unknown): config is MemoryBackendConfig => {
  if (!config || typeof config !== 'object' || config === null) return false;

  const configObj = config as Record<string, unknown>;

  if (!Object.values(MemoryBackendType).includes(configObj.backend_type as MemoryBackendType)) {
    return false;
  }

  if (configObj.backend_type === MemoryBackendType.DATABRICKS) {
    const databricksConfig = configObj.databricks_config as DatabricksMemoryConfig | undefined;
    if (!databricksConfig) return false;
    if (!databricksConfig.endpoint_name || !databricksConfig.memory_index) {
      return false;
    }
  }

  if (configObj.backend_type === MemoryBackendType.LAKEBASE) {
    const lakebaseConfig = configObj.lakebase_config as LakebaseMemoryConfig | undefined;
    if (!lakebaseConfig) return false;
    if (!lakebaseConfig.memory_table) return false;
  }

  return true;
};

/**
 * Knowledge sources (RAG) require a Lakebase pgvector memory backend, where
 * knowledge-file embeddings are stored. The default (local SQLite) backend
 * cannot store/retrieve document embeddings.
 */
export const isKnowledgeCapableMemoryConfig = (config: unknown): boolean => {
  if (!isValidMemoryBackendConfig(config)) return false;
  const c = config as MemoryBackendConfig;
  return c.backend_type === MemoryBackendType.LAKEBASE && Boolean(c.lakebase_config?.memory_table);
};

// Helper to get display name for backend type
export const getBackendDisplayName = (type: MemoryBackendType): string => {
  const displayNames: Record<MemoryBackendType, string> = {
    [MemoryBackendType.DEFAULT]: 'Local (Kasal unified Memory / SQLite)',
    [MemoryBackendType.DATABRICKS]: 'Legacy (falls back to local)',
    [MemoryBackendType.LAKEBASE]: 'Lakebase (pgvector)',
  };
  return displayNames[type] || type;
};

// Helper to get backend description
export const getBackendDescription = (type: MemoryBackendType): string => {
  const descriptions: Record<MemoryBackendType, string> = {
    [MemoryBackendType.DEFAULT]:
      'Kasal memory stored in a local SQLite database. No external infrastructure required.',
    [MemoryBackendType.DATABRICKS]:
      'Legacy configuration. Memory falls back to the local store; configure Lakebase for shared, persistent memory.',
    [MemoryBackendType.LAKEBASE]:
      'Kasal memory backed by your configured Lakebase PostgreSQL instance with pgvector. Zero additional infrastructure required.',
  };
  return descriptions[type] || '';
};

// Additional types for Databricks setup UI components
export interface EndpointInfo {
  name: string;
  type?: string;
  status?: string;
  error?: string;
  state?: string;
  ready?: boolean;
  can_delete_indexes?: boolean;
}

export interface IndexInfo {
  name: string;
  status?: string;
  index_type?: string;
}

export interface SavedConfigInfo {
  backend_id?: string;
  workspace_url?: string;
  catalog?: string;
  schema?: string;
  endpoints?: {
    memory?: EndpointInfo;
    document?: EndpointInfo;
  };
  indexes?: {
    unified?: IndexInfo;
    document?: IndexInfo;
  };
}

export interface SetupResult {
  success: boolean;
  message: string;
  endpoints?: {
    memory?: EndpointInfo;
    document?: EndpointInfo;
  };
  indexes?: {
    unified?: IndexInfo;
    document?: IndexInfo;
  };
  config?: {
    endpoint_name?: string;
    document_endpoint_name?: string;
    memory_index?: string;
    document_index?: string;
    workspace_url?: string;
    embedding_dimension?: number;
    catalog?: string;
    schema?: string;
  };
  catalog?: string;
  schema?: string;
  backend_id?: string;
  error?: string;
  warning?: string;
  info?: string;
}

export interface IndexInfoState {
  doc_count: number;
  loading: boolean;
  error?: string;
  status?: string;
  ready?: boolean;
  index_type?: string;
}

export interface ManualConfig {
  workspace_url: string;
  endpoint_name: string;
  document_endpoint_name: string;
  memory_index: string;
  document_index: string;
  embedding_model: string;
}

File: src/backend/src/schemas/memory_backend.py
"""
Memory backend configuration schemas.

Defines the Pydantic models for configuring Kasal's memory backends.
CrewAI 1.10+ uses a single unified ``Memory`` class over one storage, so these
schemas no longer split memory into short-term / long-term / entity tiers.
"""

import re as _re
from datetime import datetime
from enum import Enum
from typing import Any, Dict, Optional

from pydantic import BaseModel, Field, field_validator

# SECURITY: memory_table is interpolated into raw Lakebase SQL — restrict to a
# strict SQL identifier so it cannot inject SQL or comment out tenant filters.
_SAFE_TABLE_NAME = _re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")


class MemoryBackendType(str, Enum):
    """Supported memory backend types."""

    DEFAULT = "default"  # Local SQLite store (services/memory/local_storage_backend.py)
    DATABRICKS = "databricks"  # Databricks Vector Search
    LAKEBASE = "lakebase"  # Lakebase pgvector


class MemoryTuningConfig(BaseModel):
    """Memory Tuning knobs for a teamspace's memory (Configuration > Memory).

    The scoring weights, half-life and relevance threshold reach the storage
    backend (``MemoryBackendFactory._scoring_kwargs``); everything else is a
    declared field of the engine's ``Memory`` (``CrewMemoryService``
    forwards them), so no value set here is ever silently dropped. Any value
    left ``None`` is omitted and the code default applies.
    """

    # Composite score weights (semantic + keyword + recency + importance —
    # should roughly sum to 1.0).
    semantic_weight: Optional[float] = Field(
        None,
        ge=0.0,
        le=1.0,
        description="Weight for semantic similarity in recall (default 0.6).",
    )
    keyword_weight: Optional[float] = Field(
        None,
        ge=0.0,
        le=1.0,
        description="Weight for query-term overlap in recall (default 0.15).",
    )
    recency_weight: Optional[float] = Field(
        None,
        ge=0.0,
        le=1.0,
        description="Weight for recency decay in recall (default 0.15).",
    )
    importance_weight: Optional[float] = Field(
        None,
        ge=0.0,
        le=1.0,
        description="Weight for explicit importance in recall (default 0.1).",
    )
    recency_half_life_days: Optional[int] = Field(
        None,
        ge=1,
        description="Days for recency score to halve (default 30).",
    )
    relevance_threshold: Optional[float] = Field(
        None,
        ge=0.0,
        le=1.0,
        description=(
            "Minimum semantic similarity for a memory to be recalled at all "
            "(default 0.35). Applied before recency/importance blending, so "
            "unrelated memories are never pulled into the context."
        ),
    )
    recall_min_score: Optional[float] = Field(
        None,
        ge=0.0,
        le=1.0,
        description=(
            "Blended-score floor below which a recall returns nothing. Left "
            "unset, the floor follows the embedder in use: 0.75 with the "
            "Databricks embedder, 0.62 with the local Ollama fallback "
            "(KASAL_MEMORY_RECALL_MIN_SCORE overrides that default per "
            "deployment)."
        ),
    )

    # Consolidation.
    consolidation_threshold: Optional[float] = Field(
        None,
        ge=0.0,
        le=1.0,
        description=(
            "Similarity above which consolidation is triggered on save "
            "(default 0.85). Set 1.0 to disable consolidation."
        ),
    )
    consolidation_limit: Optional[int] = Field(
        None,
        ge=1,
        description="Max existing records compared during consolidation (default 5).",
    )
    default_importance: Optional[float] = Field(
        None,
        ge=0.0,
        le=1.0,
        description="Importance used when LLM analysis is skipped (default 0.5).",
    )

    # Recall depth control.
    confidence_threshold_high: Optional[float] = Field(
        None,
        ge=0.0,
        le=1.0,
        description="Confidence at/above which recall returns directly (default 0.8).",
    )
    confidence_threshold_low: Optional[float] = Field(
        None,
        ge=0.0,
        le=1.0,
        description="Confidence below which deeper exploration triggers (default 0.5).",
    )
    complex_query_threshold: Optional[float] = Field(
        None,
        ge=0.0,
        le=1.0,
        description="Complex-query threshold for exploration (default 0.7).",
    )
    exploration_budget: Optional[int] = Field(
        None,
        ge=0,
        description="Number of LLM-driven exploration rounds during deep recall (default 1).",
    )
    query_analysis_threshold: Optional[int] = Field(
        None,
        ge=0,
        description=(
            "Character count below which deep recall skips LLM query analysis "
            "(default 200). Set 0 to always run LLM analysis."
        ),
    )

    # LLM override for memory analysis.
    memory_llm_model: Optional[str] = Field(
        None,
        description=(
            "Override the LLM used for memory analysis (scope, importance, "
            "consolidation). Defaults to the crew's LLM so OPENAI_API_KEY "
            "is not implicitly required."
        ),
    )

    model_config = {
        "json_schema_extra": {
            "example": {
                "semantic_weight": 0.5,
                "recency_weight": 0.3,
                "importance_weight": 0.2,
                "consolidation_threshold": 0.85,
                "exploration_budget": 1,
            }
        }
    }


class DatabricksMemoryConfig(BaseModel):
    """Configuration for Databricks Vector Search as the unified memory backend."""

    # Memory endpoint (Direct Access for dynamic record-level writes).
    endpoint_name: str = Field(
        ...,
        description="Name of the Vector Search endpoint for memory (Direct Access)",
    )

    # Memory index — one index stores every MemoryRecord.
    memory_index: str = Field(
        ...,
        description=(
            "Index for Kasal memory (catalog.schema.index). Must use UNIFIED_SCHEMA."
        ),
    )

    # Document search endpoint (Storage Optimized for static corpora; unrelated to memory).
    document_endpoint_name: Optional[str] = Field(
        None,
        description=(
            "Name of the Vector Search endpoint for documents (Storage Optimized). "
            "Unrelated to memory; used by the knowledge-search feature."
        ),
    )
    document_index: Optional[str] = Field(
        None, description="Index name for document embeddings"
    )

    # Authentication (optional — environment variables / OBO also supported).
    workspace_url: Optional[str] = Field(None, description="Databricks workspace URL")
    auth_type: Optional[str] = Field(
        "default",
        description="Authentication type: default, pat, service_principal",
    )
    personal_access_token: Optional[str] = Field(
        None, description="Personal Access Token"
    )
    service_principal_client_id: Optional[str] = Field(
        None, description="Service Principal Client ID"
    )
    service_principal_client_secret: Optional[str] = Field(
        None, description="Service Principal Client Secret"
    )

    embedding_dimension: int = Field(
        1024,
        description="Dimension of embedding vectors (1024 for databricks-gte-large-en)",
    )

    # Catalog / schema metadata, used by the one-click setup flow.
    catalog: Optional[str] = Field(
        None, description="Unity Catalog name where the index is created"
    )
    schema_name: Optional[str] = Field(
        None,
        description="Schema name within catalog where the index is created",
        alias="schema",
    )

    model_config = {
        "json_schema_extra": {
            "example": {
                "endpoint_name": "vector_search_endpoint",
                "memory_index": "ml.agents.crew_memory",
                "embedding_dimension": 1024,
            }
        }
    }


class LakebaseMemoryConfig(BaseModel):
    """Configuration for Lakebase pgvector as the unified memory backend."""

    instance_name: Optional[str] = Field(
        None,
        description="Lakebase instance name (uses configured default if not set)",
    )
    embedding_dimension: int = Field(1024, description="Dimension of embedding vectors")

    # Memory table — one table stores every MemoryRecord.
    memory_table: str = Field(
        "crew_memory",
        description="Table for CrewAI memory records.",
    )

    @field_validator("memory_table")
    @classmethod
    def _validate_memory_table(cls, v: str) -> str:
        if not v or not _SAFE_TABLE_NAME.match(str(v)):
            raise ValueError(f"Invalid memory_table identifier: {v!r}")
        return v

    tables_initialized: bool = Field(
        False,
        description="Whether the memory table has been initialized on the Lakebase instance",
    )

    model_config = {
        "json_schema_extra": {
            "example": {
                "embedding_dimension": 1024,
                "memory_table": "crew_memory",
                "tables_initialized": False,
            }
        }
    }


# ---------------------------------------------------------------------------
# CRUD schemas
# ---------------------------------------------------------------------------


class MemoryBackendCreate(BaseModel):
    """Schema for creating a memory backend configuration."""

    name: str = Field(..., description="Name for this configuration")
    description: Optional[str] = Field(
        None, description="Description of the configuration"
    )
    backend_type: MemoryBackendType = Field(..., description="Type of memory backend")

    databricks_config: Optional[DatabricksMemoryConfig] = Field(None)
    lakebase_config: Optional[LakebaseMemoryConfig] = Field(None)

    cognitive_config: Optional[MemoryTuningConfig] = Field(
        None,
        description="Optional tuning parameters for the memory.",
    )

    # Escape hatch for experimental backend-specific options that don't yet
    # have first-class schema fields.
    custom_config: Optional[Dict[str, Any]] = Field(None)


class MemoryBackendUpdate(BaseModel):
    """Schema for updating a memory backend configuration."""

    name: Optional[str] = Field(None)
    description: Optional[str] = Field(None)
    backend_type: Optional[MemoryBackendType] = Field(None)

    databricks_config: Optional[DatabricksMemoryConfig] = Field(None)
    lakebase_config: Optional[LakebaseMemoryConfig] = Field(None)

    cognitive_config: Optional[MemoryTuningConfig] = Field(None)

    custom_config: Optional[Dict[str, Any]] = Field(None)
    is_active: Optional[bool] = Field(None)


class MemoryBackendResponse(BaseModel):
    """Schema for memory backend response."""

    id: str
    group_id: str
    name: str
    description: Optional[str]
    backend_type: MemoryBackendType

    databricks_config: Optional[DatabricksMemoryConfig]
    lakebase_config: Optional[LakebaseMemoryConfig]
    cognitive_config: Optional[MemoryTuningConfig]
    custom_config: Optional[Dict[str, Any]]

    is_active: bool
    is_default: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True, "arbitrary_types_allowed": True}


class MemoryBackendConfig(BaseModel):
    """Runtime configuration passed to the factory when building memory."""

    backend_type: MemoryBackendType = Field(
        MemoryBackendType.DEFAULT, description="Type of memory backend to use"
    )

    databricks_config: Optional[DatabricksMemoryConfig] = Field(
        None,
        description="Configuration for Databricks backend (required if backend_type='databricks')",
    )
    lakebase_config: Optional[LakebaseMemoryConfig] = Field(
        None,
        description="Configuration for Lakebase backend (required if backend_type='lakebase')",
    )

    cognitive_config: Optional[MemoryTuningConfig] = Field(
        None,
        description="Optional tuning parameters for the memory.",
    )

    custom_config: Optional[Dict[str, Any]] = Field(
        None, description="Additional backend-specific configuration"
    )

    model_config = {
        "json_schema_extra": {
            "example": {
                "backend_type": "databricks",
                "databricks_config": {
                    "endpoint_name": "vector_search_endpoint",
                    "memory_index": "ml.agents.crew_memory",
                    "embedding_dimension": 1024,
                },
                "cognitive_config": {
                    "semantic_weight": 0.5,
                    "recency_weight": 0.3,
                    "importance_weight": 0.2,
                },
            }
        }
    }
